### DISCUSSION TITLE
Count the number of results files on the server.

### DISCUSSION THREAD

**Discussion by holgerMitterer** [jspsych/jsPsych#3248] [Q&A]
We are running an experiment from populations that are not available on Prolific or similar sites, so we let participants run online and they can collect their reward at specific times in a specific place. The problem is that we want to stop the experiment if there are sufficient participants (and we can't run more because of we need to pay participants). Experiments are run on parts of the uni-server we have access to. The idea is to count the number of csv files in the data directory on the server, and if that is larger than (e.g.) 40, the timeline is changed to a single trial saying "Sorry, this experiment has gathered the required number of participants".  However, I struggle to find a solution that simply counts the number of csv files on the server (i.e., in the subdir ./data to which csv files are written at the end of the study). 

The solutions I find online do not work (since they require node.js, it is unclear to me how to get that working on the server), and my lack of familiarity with online programming make it difficult for me to understand what the best solution here is.

Any suggestions?

Kind regards,
Holger

**Comment by jodeleeuw** [comment]
Hi @holgerMitterer,

This is going to require some kind of server-side program to run, like node or php. Client-side JS like jsPsych can't do this because it would be a security breach of the server. Do you know if PHP is installed on the server? That might be the easiest way to do it.

**Reply by holgerMitterer** [reply]
Yes, php is running on the server we are using, that is how we save the csv files.

**Reply by jodeleeuw** [reply]
Writing a php script to count the number of files in the directory is pretty easy. See, for example, [this Stack Overflow post](https://stackoverflow.com/questions/12801370/count-how-many-files-in-directory-php). 

From there you could either just make the experiment .html a .php file instead and display a message instead of loading the jsPsych experiment, or you could use the JavaScript fetch API to call the PHP file and then use the result to update the experiment.

Here're resources for option 1: [blog post](https://www.geeksforgeeks.org/how-to-use-php-in-html/)

Here're resources for option 2: [stack overflow post](https://stackoverflow.com/questions/71442633/how-to-send-request-to-php-file-via-fetch-api-post-method)

**Reply by holgerMitterer** [reply]
Thanks for these pointers, I got it solved and document it here for further use:

I generated the follwoing php file (fileCounter.php)

`<?php
       $directory = "./data/";
      $nOfFiles = count(glob($directory . "*.csv"));
      header('Content-Type: application/json');
      echo json_encode($nOfFiles);
?>`
 

And within the jsPsych script added the following:

  //generating global variable
  var numberOfCSVfiles;
  //calling the php
  function countFiles() {
    fetch('fileCounter.php')
      .then(response => {        return response.json(); })
      .then(data => {
        // Use the data from the PHP script
        console.log(data);
        // and saving it in the global variable
        numberOfCSVfiles = data;
      })
  }

  var callFileCounter = {
    type: 'call-function',
    func: countFiles
  }

And the I used this in a trial as follows:
  var payTrial = {
	    type: 'html-keyboard-response',
      stimulus:  function () {
        var dispaly = "";
        console.log('number of CSV files: ', numberOfCSVfiles)
			     if (numberOfCSVfiles < 41)
			        {
   		             display = '<p>You are within the first 40 participants, you can proceed with the experiment.<br>Press a key to continue</p>';
			              }else
			        { display = '<p>More than 40 participants already participated. You can proceed, but we do not have funds to pay you anymore.</p>\
                   if you want to stop, simply close this window, otherwise, press any key.'}
			     return display;
		   },
		data: {},
		choices: jsPsych.ALL_KEYS,
		trial_duration: 1000000
	}


Now one odd thing, this only worked if another trial intervened in the timeline.
so this worked:
	timeline.push(callFileCounter, startOfExperiment1, payTrial)
but with this one, it didn't
	timeline.push(callFileCounter,  payTrial, startOfExperiment1)

Because payTrial did not yet have the updated value of the global variable "numberOfCSV files"
The console.log check up there then return the following:

     number of CSV files:  undefined

But at least, I got it to work.

Thanks for the help,
Holger 

PS: Sorry, tested this on an old version 6 script, sorry for the old syntax.
